src/Agent.Listener/Environment.cs

```diff
@@ -0,0 +1,338 @@
+                }
+
+                token.ThrowIfCancellationRequested();
+            }
+        }
+
+        private void GetFilteredEnvironmentVars(Dictionary<string, string> vars)
+        {
+            IDictionary envVars = System.Environment.GetEnvironmentVariables();
+
+            // Begin with ignoring env vars declared herein
+            var ignoredEnvVariables = new HashSet<string>(_ignoredEnvVariables);
+
+            // Also ignore env vars specified in the 'VSO_AGENT_IGNORE' env var
+            if (envVars.Contains(EnvIgnore))
+            {
+                var additionalIgnoredVars = ((string)envVars[EnvIgnore]).Split(',');
+                foreach (var ignored in additionalIgnoredVars)
+                {
+                    ignoredEnvVariables.Add(ignored.Trim());
+                }
+            }
+
+            // Get filtered env vars
+            foreach (DictionaryEntry envVar in envVars)
+            {
+                string varName = (string)envVar.Key;
+                string varValue = (string)envVar.Value ?? string.Empty;
+                if (!ignoredEnvVariables.Contains(varName) && varValue.Length < 1024)
+                {
+                    vars.Add(varName, varValue);
+                }
+            }
+        }
+
+        // Ensures existence of the environment file at the specified path, creating it if missing.
+        public void EnsureEnvFile()
+        {
+            string envFileName = IOUtil.GetEnvFilePath();
+            if (File.Exists(envFileName))
+            {
+                return;
+            }
+
+            var envVars = new Dictionary<string, string>();
+            GetFilteredEnvironmentVars(envVars);
+
+            var content = new StringBuilder(string.Empty);
+            foreach (var envvar in envVars)
+            {
+                content.Append($"{envvar.Key}={envvar.Value}");
+                content.AppendLine();
+            }
+
+            File.WriteAllText(envFileName, content.ToString(), System.Text.Encoding.UTF8);
+        }
+
+        // Gets the environment that the agent and worker will use when running as a service.
+        // The current process' environment is overlayed with contents of the environment file.
+        // When not running as a service, the interactive/shell process' environment is used.
+        public IDictionary<string, string> GetEnv()
+        {
+            string envFileName = IOUtil.GetEnvFilePath();
+            if (!File.Exists(envFileName))
+            {
+                throw new FileNotFoundException();
+            }
+
+            string data = File.ReadAllText(envFileName, System.Text.Encoding.UTF8);
+
+            var env = new Dictionary<string, string>();
+
+            var lines = data.Split(new string[] { System.Environment.NewLine }, StringSplitOptions.RemoveEmptyEntries);
+            foreach (string line in lines)
+            {
+                var tokens = line.Split('=');
+                if (tokens.Length == 2)
```

Comment:
It seems strange to restrict values that contain '=' or newlines (main concerns). These restrictions can be lifted via escaping, e.g. json. There's less common restrictions that could be avoided that way as well ('=' in the key itself), just by using json. Bryan may have some thoughts here.